tests/unit/api/apiUtils/objectLockHelpers.js

```diff
@@ -0,0 +1,169 @@
+        const headers = {
+            'x-amz-object-lock-mode': 'GOVERNANCE',
+        };
+        const objectLockValidation = validateHeaders(bucketInfo, headers, log);
+        const expectedError = errors.InvalidArgument.customizeDescription(
+            'x-amz-object-lock-retain-until-date and x-amz-object-lock-mode ' +
+            'must both be supplied');
+        assert.strictEqual(objectLockValidation.error.InvalidArgument, true);
+        assert.strictEqual(objectLockValidation.error.description,
+            expectedError.description);
+    });
+
+    it('should pass with past retention date header', () => {
+        const headers = {
+            'x-amz-object-lock-retain-until-date': '2005-10-12',
+            'x-amz-object-lock-mode': 'COMPLIANCE',
+        };
+        const expectedError = errors.InvalidArgument.customizeDescription(
+            'The retain until date must be in the future!')
+        const objectLockValidation = validateHeaders(bucketInfo, headers, log);
+        assert.strictEqual(objectLockValidation.error.InvalidArgument, true);
+        assert.strictEqual(objectLockValidation.error.description,
+            expectedError.description);
+    });
+
+    it('should fail with invalid legal hold header', () => {
+        const headers = {
+            'x-amz-object-lock-legal-hold': 'on',
+        };
+        const objectLockValidation = validateHeaders(bucketInfo, headers, log);
+        const expectedError = errors.InvalidArgument.customizeDescription(
+            'Unknown wormMode directive');
+        assert.strictEqual(objectLockValidation.error.InvalidArgument, true);
+        assert.strictEqual(objectLockValidation.error.description,
+            expectedError.description);
+    });
+
+    it('should pass with invalid retention period header', () => {
+        const headers = {
+            'x-amz-object-lock-retain-until-date': '2050-10-12',
+            'x-amz-object-lock-mode': 'Governance',
+        };
+        const objectLockValidation = validateHeaders(bucketInfo, headers, log);
+        const expectedError = errors.InvalidArgument.customizeDescription(
+            'Unknown wormMode directive');
+        assert.strictEqual(objectLockValidation.error.InvalidArgument, true);
+        assert.strictEqual(objectLockValidation.error.description,
+            expectedError.description);
+    });
+});
+
+describe('objectLockHelpers: calculateRetainUntilDate', () => {
+    it('should calculate retainUntilDate for config with days', () => {
+        const mockConfigWithDays = {
+            mode: 'GOVERNANCE',
+            days: 90,
+        };
+        const date = new Date();
+        const expectedRetainUntilDate
+            = date.setDate(date.getDate() + mockConfigWithDays.days);
+        const retainUntilDate = calculateRetainUntilDate(mockConfigWithDays);
+        assert.strictEqual(retainUntilDate.toString().slice(0, 24),
```

Comment:
instead of to string + slice you could use a fixed date above and with your expected.

Reply:
I don't think I could cause `calculateRetainUntilDate` calculates the date based on the current date which is why I needed to mimic the same logic (unless I misunderstood your comment 🤔 ).

Reply:
oooh, I see, yea this would be much easier to test if `calculateRetainUntilDate` was a pure method and took in a date value. Because it has the side effect of getting the current date on its own, there will be nanosecond variations between the one calculated in the method and our test. Alternatively, you could use Sinon to mock the moment call to a specific date but pure functions are great regardless.

Reply:
hmm actually I like the idea of using Sinon to mock the moment call to a specific date; not sure how I never thought of Sinon here at all lol (weird cause I'm used to using it a lot). Might be bit tight for this PR but I'll have some follow-up refactors (such as adding custom errors etc.) anyway so gonna give this a try to see if it's better mocked instead.

Reply:
Sounds good to me, resolve this if you are going to push it back.